File: chapter-3_numerical-algorithms/python/sieveEras/sieve_eras.py
def sieveEras(a, b):
    numList = list(range(a, b+1))
    #print(numList)

    firstTwoIndex = None
    count = 0
    while firstTwoIndex is None:
        currentNumber = numList[count]
        if (currentNumber != 2) and ((currentNumber % 2) == 0):
            firstTwoIndex = count
            numList[count] = None
        count += 1
    #print(numList)

    count = firstTwoIndex + 2
    while count < len(numList):
        numList[count] = None
        count += 2

    if numList[0] == 1:
        numList[0] = None
    #   print(numList)

    count = firstTwoIndex + 1
    while count < len(numList):
        #print(numList[count])
        if not numList[count] is None:
            if is_prime(numList[count]):
                #print(numList[count])
                multiple = numList[count] * numList[count]
                while multiple <= len(numList):
                    #print(multiple)
                    numList[multiple] = None
                    multiple += 2 * count
            else:
                #print(numList[count])
                numList[count] = None
        count += 1

    result = []
    count = 0
    while count < len(numList):
        if not (numList[count] is None):
            result.append(numList[count])
        count += 1
    print('Eratosthenes found {} prime numbers between {} and {}'.format(len(result), a, b));
    return (result)


def is_prime(n):
    if n == 2 or n == 3: return True
    if n < 2 or n%2 == 0: return False
    if n < 9: return True
    if n%3 == 0: return False
    r = int(n**0.5)
    f = 5
    while f <= r:
        if n%f == 0: return False
        if n%(f+2) == 0: return False
        f +=6
    return True

File: chapter-3_numerical-algorithms/javascript/sieveEras/sieve-eras.js
'use strict';

var isPrime = require('../fermatPrime/fermat-prime').isPrime;

//Return an array of the prime numbers between the two numbers:
function sieve(a, b) {
  a = a || 0;
  b = b || 0;
  if(b < a) {
    throw new Error('b must be larger than or equal to a');
  }
  var arrayNumbers = [];
  //O(N)
  if(a === 1) {
    a++;
  }
  for (var i = a; i <= b; i++) {
    arrayNumbers.push(i);
  }
  //console.log(arrayNumbers.length);
  //console.log(JSON.stringify(arrayNumbers));
  //find first multiple of 2, if it's 2, then leave it be and start looping, if it's not, then :
  var firstTwoIndex;
  var count = 0;
  while(!firstTwoIndex && firstTwoIndex !== 0) {
    var currentNum = arrayNumbers[count];
    if((Math.abs(currentNum % 2) === 0) && currentNum != 2) {
      //cross that number out:
      //console.log(arrayNumbers[count]);
      arrayNumbers[count] = null;
      firstTwoIndex = count;
    } else {
      count++;
    }
  }
  //console.log(JSON.stringify(arrayNumbers));
  //console.log(firstTwoIndex);
  //Remove all multiples of 2 for we know those are not primes:
  for (var i = firstTwoIndex; i < arrayNumbers.length; i += 2) {
    arrayNumbers[i] = null;
  }
  //console.log(JSON.stringify(arrayNumbers));
  //console.log(JSON.stringify(arrayNumbers));

  for (var i = firstTwoIndex + 1; i < arrayNumbers.length; i++) {
    //console.log(i + '? ' + isPrime(i));
    if(isPrime(arrayNumbers[i])) {
      //console.log(arrayNumbers[i] + ' is prime');
      for (var multiple = (arrayNumbers[i] * arrayNumbers[i]); multiple <= arrayNumbers.length; multiple += 2 * arrayNumbers[i]) {
        //console.log(multiple +': ' + arrayNumbers[multiple+1]);
        arrayNumbers[multiple] = null;
      }
    } else {
      arrayNumbers[i] = null;
    }
  }
  var result = [];
  for (var i = 0; i < arrayNumbers.length; i++) {
    if(arrayNumbers[i] !== null) {
      result.push(arrayNumbers[i]);
    }
  }
  console.log('Eratosthenes found ' + result.length + ' prime numbers between ' + (a-1) + ' and ' + b);
  return result;
}

module.exports = sieve;
